### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by juampynr**
Drupal.org issue: https://www.drupal.org/node/2913320

This is a work in progress. I want to get coverage up to 70%.

image

### CODE REVIEW

**Comment by juampynr** [comment]
4%. This will take me longer as there is more code.

image

**Comment by juampynr** [comment]
@nignatov I just discovered [this test file in the module](https://github.com/nodespark/elasticsearch_connector/blob/8.x-5.x/src/Tests/ElasticsearchTest.php). How does it work?

I also created https://www.drupal.org/node/2913320 based on https://github.com/nodespark/elasticsearch_connector/pull/19#issuecomment-333529276.

**Comment by juampynr** [comment]
@nignatov, this is as far as I could get with Unit tests. The rest of the code needs some refactoring so we can test it via unit tests and then the remaining bit will be tested via Kernel tests. Finally, I suggest that we test the UI with Behat tests through drupalextension, which we can run here in PRs through CircleCI.

**Comment by juampynr** [comment]
@nignatov, can you have a look at this when you get a chance? I am already working on a separate pull request where I am moving some static methods into services so they can be unit tested.

**Comment by juampynr** [comment]
I have removed a hardcoded dependency by making ClusterFactory into a service. Code coverage is now at 8%:

image

By transforming the remaining factories to services we will be able to add further testing. There is also some code that I don't see value in testing it with unit tests like forms and controllers. Instead, I have added CircleCI integration so we can write Behat tests for them.

**Comment by juampynr** [comment]
@nignatov, I enabled CircleCI integration with a [sample PR in my fork](https://github.com/juampynr/elasticsearch_connector/pull/1) and here is what I get so far:

image

The good thing about CircleCI is that it gives us full freedom to test in a container provided by us. I plan to add Behat tests where we can test the whole interaction set that happens between Drupal and an Elasticsearch server.

If you are interested in incorporating CircleCI support for this project, just go to https://circleci.com/dashboard, choose this repository and enable it.

**Comment by nignatov** [comment]
Sounds good to me. I will check this tomorrow. Thank you @juampynr

File: modules/elasticsearch_connector_views/src/Plugin/views/filter/ElasticsearchViewsStringFilter.php [hunk 1]
```diff
@@ -191,7 +191,9 @@ public function operators() {
    * Helper function to query.
    */
   public function query() {
-
+    if (!empty($this->value[0])) {
```
Comment (m4olivei) [inline modules/elasticsearch_connector_views/src/Plugin/views/filter/ElasticsearchViewsStringFilter.php:194]: What is this for?

File: modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php [hunk 2]
```diff
@@ -124,6 +133,9 @@ public static function create(ContainerInterface $container, array $configuratio
                         ->get('elasticsearch_connector_views');
     $plugin->setLogger($logger);
 
+    $entity_type_manager = $container->get('entity_type.manager');
+    $plugin->setEntityTypeManager($entity_type_manager);
```
Comment (m4olivei) [inline modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php:137]: This is sort of atypical, why add and use a setter, why not do the usual thing with a create method, which is to implement a constructor, and pass the services/arguments to set as members of the class to it?

File: modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php [hunk 3]
```diff
@@ -300,25 +323,6 @@ public function build(ViewExecutable $view) {
     $this->params['fields'][] = '_source';
 
     $params = array();
-    if (isset($this->params['q']) && !empty($this->params['q'])) {
```
Comment (m4olivei) [inline modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php:303]: Perhaps this is out of scope? Can you elaborate on what's being removed here?

File: src/ElasticSearch/Parameters/Factory/MappingFactory.php [hunk 4]
```diff
@@ -16,66 +16,62 @@ class MappingFactory {
...
+        return [
+          'type' => 'text',
+          'boost' => $field->getBoost(),
+          'fields' => [
+            "keyword" => [
+              "type" => 'keyword',
+              'ignore_above' => 256,
             ]
-          ];
+          ]
```
Comment (m4olivei) [inline src/ElasticSearch/Parameters/Factory/MappingFactory.php:34]: Nit, missing trailing comma.

File: tests/src/Unit/ElasticSearch/Parameters/Builder/SearchBuilderTest.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+   */
+  public function testConstruct() {
+    $this->assertInstanceOf(SearchBuilder::class, $this->searchBuilder);
+  }
+
+  /**
+   * @covers ::build
+   */
+  public function testBuild() {
+    // TODO Can't test because IndexFactory is hardcoded
```
Comment (m4olivei) [inline tests/src/Unit/ElasticSearch/Parameters/Builder/SearchBuilderTest.php:49]: Follow up issue?

File: tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    $filter = FilterFactory::filterFromCondition($condition->reveal());
+    $expected_filter = [
+      'exists' =>
+        [
+          'field' => 'foo',
+        ],
+    ];
+    $this->assertEquals($expected_filter, $filter);
+
+    // Thest the = operator.
```
Comment (m4olivei) [inline tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php:46]: Typo on "Thest"

File: tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    FilterFactory::filterFromCondition($condition->reveal());
+  }
+
+  /**
+   * @covers ::filterFromCondition
+   */
+  public function testFilterFromConditionB() {
+    // Normal filters
+    /** @var \Prophecy\Prophecy\ObjectProphecy $condition */
+    $condition = $this->prophesize(Condition::class);
```
Comment (m4olivei) [inline tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php:92]: The pattern of tests here is fine, but it seems like a good fit for a `@dataProvider`, where for each test case it would pass the value, operator, field name and expected result.

File: tests/src/Unit/ElasticSearch/Parameters/Factory/MappingFactoryTest.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    $field->getType()
+      ->willReturn('text');
+    $field->getBoost()
+      ->willReturn(1);
+
+    $expected_mapping = [
+      'type' => 'text',
+      'boost' => 1,
+      'fields' => [
+        "keyword" => [
```
Comment (m4olivei) [inline tests/src/Unit/ElasticSearch/Parameters/Factory/MappingFactoryTest.php:31]: Nit, inconsistency in use of quotations.

**Review by m4olivei — APPROVED** [review]
This looks great! Thanks for adding the tests and CI. I added a couple comments along the way, but nothing blocking merge.

**Comment by nignatov** [comment]
@juampynr I've received the following error in Circles when merged your changes:

"Build-agent version 0.0.4355-1897c1d (2017-10-13T22:25:17+0000)
Configuration errors: 1 error occurred:

* Cannot find a job named `build` to run in the `jobs:` section of your configuration file.
If you expected a workflow to run, check your config contains a top-level key called 'workflows:'"

Do I miss something or it is an issue?

**Comment by juampynr** [comment]
Can you post the link there to the failure? There is still some work to do to get all jobs working so I will continue with that and create a new pull request.

**Comment by nignatov** [comment]
Sure, here you go: https://circleci.com/gh/nodespark/elasticsearch_connector/1?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: modules/elasticsearch_connector_views/src/Plugin/views/filter/ElasticsearchViewsStringFilter.php (hunk 1 above)
Comment (juampynr) [reply modules/elasticsearch_connector_views/src/Plugin/views/filter/ElasticsearchViewsStringFilter.php:194]: Phew, I honestly can't remember!

File: modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php (hunk 2 above)
Comment (juampynr) [reply modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php:137]: I know, but this plugin is a bit strange and I did not want to change the architecture while implementing tests so I followed what was there already to add the dependency.

Comment (juampynr) [reply modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php:137]: We could come back to this in a new PR and use the constructor to inject dependencies.

File: modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php (hunk 3 above)
Comment (juampynr) [reply modules/elasticsearch_connector_views/src/Plugin/views/query/ElasticsearchViewsQuery.php:303]: This must have been the result of merging the base branch on top.

File: tests/src/Unit/ElasticSearch/Parameters/Builder/SearchBuilderTest.php (hunk 5 above)
Comment (juampynr) [reply tests/src/Unit/ElasticSearch/Parameters/Builder/SearchBuilderTest.php:49]: Yep, I will create it.

File: tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php (hunk 6 above)
Comment (juampynr) [reply tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php:46]: Oops! Will fix that in a follow up.

File: tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php (hunk 7 above)
Comment (juampynr) [reply tests/src/Unit/ElasticSearch/Parameters/Factory/FilterFactoryTest.php:92]: Yup, I thought that it was clear this way but there is a lot of code duplicity. I will change it into a data provider.
